purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt

```diff
@@ -0,0 +1,118 @@
+package com.revenuecat.purchases
+
+import android.content.Context
+import android.preference.PreferenceManager
+import androidx.test.ext.junit.rules.activityScenarioRule
+import com.revenuecat.purchases.common.BillingAbstract
+import com.revenuecat.purchases.models.StoreTransaction
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.slot
+import org.assertj.core.api.Assertions
+import org.junit.Rule
+import java.net.URL
+import java.util.Date
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration.Companion.seconds
+
+open class BasePurchasesIntegrationTest {
+
+    @get:Rule
+    var activityScenarioRule = activityScenarioRule<MainActivity>()
+
+    protected val testTimeout = 5.seconds
+    protected val currentTimestamp = Date().time
+    protected val testUserId = "android-integration-test-$currentTimestamp"
+    protected val proxyUrl = Constants.proxyUrl.takeIf { it != "NO_PROXY_URL" }
+
+    protected lateinit var mockBillingAbstract: BillingAbstract
+
+    protected var latestPurchasesUpdatedListener: BillingAbstract.PurchasesUpdatedListener? = null
+    protected var latestStateListener: BillingAbstract.StateListener? = null
+
+    // region helpers
+
+    protected fun setupTest(
+        initialSharedPreferences: Map<String, String> = emptyMap(),
+        initialActivePurchases: Map<String, StoreTransaction> = emptyMap(),
+        postSetupTestCallback: (MainActivity) -> Unit = {}
+    ) {
+        latestPurchasesUpdatedListener = null
+        latestStateListener = null
+
+        onActivityReady {
+            clearAllSharedPreferences(it)
+            writeSharedPreferences(it, initialSharedPreferences)
+
+            mockBillingAbstract = mockk<BillingAbstract>(relaxed = true).apply {
+                every { purchasesUpdatedListener = any() } answers { latestPurchasesUpdatedListener = firstArg() }
+                every { stateListener = any() } answers { latestStateListener = firstArg() }
+            }
+            mockActivePurchases(initialActivePurchases)
+
+            proxyUrl?.let { urlString ->
+                Purchases.proxyURL = URL(urlString)
+            }
+
+            Purchases.configure(
+                PurchasesConfiguration.Builder(it, Constants.apiKey)
+                    .appUserID(testUserId)
+                    .build(),
+                mockBillingAbstract
+            )
+
+            postSetupTestCallback(it)
+        }
+    }
+
+    protected fun tearDownTest() {
+        Purchases.sharedInstance.close()
+    }
+
+    protected fun onActivityReady(block: (MainActivity) -> Unit) {
+        activityScenarioRule.scenario.onActivity(block)
+    }
+
+    protected fun ensureBlockFinishes(block: (CountDownLatch) -> Unit) {
```

Comment:
Copied this from the `BackendIntegrationTest` which are actually unit tests unlike this.